db/change_cache.go

```diff
@@ -265,48 +267,64 @@ func (c *changeCache) CleanAgedItems() {
+// and subsequent removal (RemoveSkipped).
 func (c *changeCache) CleanSkippedSequenceQueue() {
 
-	foundEntries, pendingDeletes := func() ([]*LogEntry, []uint64) {
-		c.skippedSeqLock.Lock()
-		defer c.skippedSeqLock.Unlock()
-
-		var found []*LogEntry
-		var deletes []uint64
-		for _, skippedSeq := range c.skippedSeqs {
-			if time.Since(skippedSeq.timeAdded) > c.options.CacheSkippedSeqMaxWait {
-				// Attempt to retrieve the sequence from the view before we remove.
-				startSequence := skippedSeq.seq
-				endSequence := skippedSeq.seq
-				// Note: The view query is only going to hit for active revisions - sequences associated with inactive revisions
-				//       aren't indexed by the channel view.  This means we can potentially miss channel removals:
-				//       when an older revision is missed by the TAP feed, and a channel is removed in that revision,
-				//       the doc won't be flagged as removed from that channel in the in-memory channel cache.
-				entries, err := c.context.getChangesInChannelFromQuery("*", startSequence, endSequence, 0, false)
-				if err == nil && len(entries) > 0 {
-					// Found it - store to send to the caches.
-					found = append(found, entries[0])
-				} else {
-					if err != nil {
-						base.Warnf(base.KeyAll, "Error retrieving changes from view during skipped sequence check: %v", err)
-					}
-					base.Warnf(base.KeyAll, "Skipped Sequence %d didn't show up in MaxChannelLogMissingWaitTime, and isn't available from the * channel view.  If it's a valid sequence, it won't be replicated until Sync Gateway is restarted.", skippedSeq.seq)
-				}
-				// Remove from skipped queue
-				deletes = append(deletes, skippedSeq.seq)
-			} else {
-				// skippedSeqs are ordered by arrival time, so can stop iterating once we find one
-				// still inside the time window
-				break
+	oldSkippedSequences := c.GetSkippedSequencesOlderThanMaxWait()
+	var foundEntries []*LogEntry
+	var pendingRemovals []uint64
+
+	if len(oldSkippedSequences) == 0 {
+		return
+	}
+
+	base.Infof(base.KeyCache, "Starting CleanSkippedSequenceQueue, found %d skipped sequences older than max wait for database %s", len(oldSkippedSequences), base.UD(c.context.Name))
+
+	if c.context.Options.UnsupportedOptions.DisableCleanSkippedQuery == true {
+		pendingRemovals = append(pendingRemovals, oldSkippedSequences...)
+		oldSkippedSequences = nil
+	}
+
+	for len(oldSkippedSequences) > 0 {
+		var skippedSeqBatch []uint64
+		if len(oldSkippedSequences) >= SkippedSeqCleanViewBatch {
+			skippedSeqBatch = oldSkippedSequences[0:SkippedSeqCleanViewBatch]
+			oldSkippedSequences = oldSkippedSequences[SkippedSeqCleanViewBatch:]
+		} else {
+			skippedSeqBatch = oldSkippedSequences[0:]
+			oldSkippedSequences = nil
+		}
+
+		base.Debugf(base.KeyCache, "Issuing skipped sequence clean query for %d sequences, %d remain pending (db:%s).", len(skippedSeqBatch), len(oldSkippedSequences), base.UD(c.context.Name))
```

Comment:
See other comments related to log levels.

```suggestion
		base.Infof(base.KeyCache, "Issuing skipped sequence clean query for %d sequences, %d remain pending (db:%s).", len(skippedSeqBatch), len(oldSkippedSequences), base.UD(c.context.Name))
```